core/src/http_helpers.rs

```diff
@@ -27,18 +27,25 @@
 //! Utility methods relying on hyper
 
 use crate::error::GenericTransportError;
-use futures_util::stream::StreamExt;
+use anyhow::anyhow;
+use hyper::body::{Buf, HttpBody};
+use std::error::Error as StdError;
 
-/// Read a data from a [`hyper::Body`] and return the data if it is valid and within the allowed size range.
+/// Read a data from [`hyper::body::HttpBody`] and return the data if it is valid JSON and within the allowed size range.
 ///
 /// Returns `Ok((bytes, single))` if the body was in valid size range; and a bool indicating whether the JSON-RPC
 /// request is a single or a batch.
 /// Returns `Err` if the body was too large or the body couldn't be read.
-pub async fn read_body(
+pub async fn read_body<B>(
 	headers: &hyper::HeaderMap,
-	mut body: hyper::Body,
+	body: B,
 	max_request_body_size: u32,
-) -> Result<(Vec<u8>, bool), GenericTransportError<hyper::Error>> {
+) -> Result<(Vec<u8>, bool), GenericTransportError>
+where
+	B: HttpBody + Send + 'static,
+	B::Data: Send,
+	B::Error: Into<Box<dyn StdError + Send + Sync>>,
+{
 	// NOTE(niklasad1): Values bigger than `u32::MAX` will be turned into zero here. This is unlikely to occur in
 	// practice and for that case we fallback to allocating in the while-loop below instead of pre-allocating.
 	let body_size = read_header_content_length(headers).unwrap_or(0);
```

Comment:
Just a nit: what if somebody sets a `Content-Length: u32::MAX` type header just to DoS this with huge allocations?

Maybe it's best to do a `.max(1024 * 1024 * 4)` or something to set a sane limit on how much to pre-allocate?

Reply:
(could also take `max_request_body_size` into account?)

Reply:
we have this check before allocating:

```rust
	if body_size > max_request_body_size {
		return Err(GenericTransportError::TooLarge);
	}
```

Thus, we can't allocate more than `max_request_body_size`. Probably sufficient?

Reply:
we could take it one stop further though as you said with `min(max_request_body_size, 1024 * 16)`

Reply:
Yeah maybe both is good; that way something can't force us to pre-allocate `max_body_size` every time even unless it actually sends that much data! But with that check you mentioned in place this is def much less important so whatever you think best :)

Reply:
that's probably fair, done